From 91f6f53f59d3a451ab4e423f42c7c6cdf3800984 Mon Sep 17 00:00:00 2001
From: Matthew <matthew@urgero.net>
Date: Fri, 15 Dec 2023 12:50:56 -0600
Subject: [PATCH] Add activation code for Joshua upgrade (#310)

Make all necessary preparations for the Joshua upgrade Dec 22, 2023.

Used commit 6e2e115 as a template for this commit.

---------
Co-authored-by: Shammah Chancellor <213740+schancel@users.noreply.github.com>

Rebase provenance: this historical change was adapted onto the pinned
Bitcoin ABC base. Source diff is the recorded Lotus V1 replay commit below.

Historical-source-commit: 6d7f6ad49192fd570fd13a6c554681885efca291
Recorded-replay-commit: f46039ca31053424d7b33d77f522f1c9eb47df95

diff --git a/src/consensus/activation.cpp b/src/consensus/activation.cpp
index c8df6ddd48..d3e30b598d 100644
--- a/src/consensus/activation.cpp
+++ b/src/consensus/activation.cpp
@@ -92,3 +92,13 @@ bool IsDeuteronomyEnabled(const Consensus::Params &params,
     return pindexPrev->GetMedianTimePast() >=
            gArgs.GetIntArg("-deuteronomyactivationtime", params.deuteronomyActivationTime);
 }
+
+bool IsJoshuaEnabled(const Consensus::Params &params,
+                     const CBlockIndex *pindexPrev) {
+    if (pindexPrev == nullptr) {
+        return false;
+    }
+
+    return pindexPrev->GetMedianTimePast() >=
+           gArgs.GetIntArg("-joshuaactivationtime", params.joshuaActivationTime);
+}
diff --git a/src/consensus/activation.h b/src/consensus/activation.h
index 29aa6fbe3c..d29330a544 100644
--- a/src/consensus/activation.h
+++ b/src/consensus/activation.h
@@ -42,4 +42,8 @@ bool IsNumbersEnabled(const Consensus::Params &params,
 bool IsDeuteronomyEnabled(const Consensus::Params &params,
                           const CBlockIndex *pindexPrev);
 
+/** Check if 2023-12-22T03:27:00.000Z protocol upgrade has activated. */
+bool IsJoshuaEnabled(const Consensus::Params &params,
+                     const CBlockIndex *pindexPrev);
+
 #endif // BITCOIN_CONSENSUS_ACTIVATION_H
diff --git a/src/consensus/addresses.h b/src/consensus/addresses.h
index a265d305d7..a971ed6f76 100644
--- a/src/consensus/addresses.h
+++ b/src/consensus/addresses.h
@@ -133,5 +133,9 @@ Consensus::CoinbaseAddresses AddressSets = {
             "lotus_16PSJHA7AxHkiV4XACtMRSzKmw8F1pT3ZbsEnzJ4s",
             "lotus_16PSJHA7AxHkiV4XACtMRSzKmw8F1pT3ZbsEnzJ4s",
         },
+    .joshua =
+        {
+            "lotus_16PSJJYp3YkVyctW8LRbp6UhbsrQMJghL3jjze3b7",
+        },
 };
 } // namespace RewardAddresses
diff --git a/src/consensus/params.h b/src/consensus/params.h
index e08be475bf..a1027f696f 100644
--- a/src/consensus/params.h
+++ b/src/consensus/params.h
@@ -23,6 +23,7 @@ struct CoinbaseAddresses {
     std::vector<std::string> leviticus;
     std::vector<std::string> numbers;
     std::vector<std::string> deuteronomy;
+    std::vector<std::string> joshua;
 };
 
 enum BuriedDeployment : int16_t {
@@ -58,6 +59,9 @@ struct Params {
     /** Unix time used for MTP activation of 2023-12-22T03:27:00.000Z protocol
      * upgrade */
     int joshuaActivationTime;
+    /** Unix time used for MTP activation of 2024-06-20T20:51:00.000Z protocol
+     * upgrade */
+    int judgesActivationTime;
     /** Block height at which BIP16 becomes active */
     int BIP16Height;
     /** Block height and hash at which BIP34 becomes active */
diff --git a/src/init.cpp b/src/init.cpp
index 9ba88c59e2..ebc0dd3f20 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -472,7 +472,7 @@ void SetupServerArgs(NodeContext &node) {
         "-splash",
         "-uiplatform",
         // TODO remove after the May 2026 upgrade
-        "-obolenskyactivationtime", "-exodusactivationtime", "-leviticusactivationtime", "-numbersactivationtime", "-deuteronomyactivationtime",
+        "-obolenskyactivationtime", "-exodusactivationtime", "-leviticusactivationtime", "-numbersactivationtime", "-deuteronomyactivationtime", "-joshuaactivationtime",
     };
 
     // Set all of the args and their help
diff --git a/src/kernel/chainparams.cpp b/src/kernel/chainparams.cpp
index c4b031add6..78a74e245e 100644
--- a/src/kernel/chainparams.cpp
+++ b/src/kernel/chainparams.cpp
@@ -130,6 +130,8 @@ public:
         consensus.deuteronomyActivationTime = 1687359480;
         // 2023-12-22T03:27:00.000Z protocol upgrade
         consensus.joshuaActivationTime = 1703215620;
+        // 2024-06-20T20:51:00.000Z protocol upgrade
+        consensus.judgesActivationTime = 1718916600;
 
         /**
          * The message start string is designed to be unlikely to occur in
@@ -309,6 +311,9 @@ public:
         // 2023-12-22T03:27:00.000Z protocol upgrade
         consensus.joshuaActivationTime =
             mainnetConsensus.joshuaActivationTime - testnetActivationOffset;
+        // 2024-06-20T20:51:00.000Z protocol upgrade
+        consensus.judgesActivationTime =
+            mainnetConsensus.judgesActivationTime - testnetActivationOffset;
 
         // "ltdk" with MSB set
         diskMagic[0] = 0xec;
@@ -459,6 +464,8 @@ public:
             mainnetConsensus.deuteronomyActivationTime;
         // 2023-12-22T03:27:00.000Z protocol upgrade
         consensus.joshuaActivationTime = mainnetConsensus.joshuaActivationTime;
+        // 2024-06-20T20:51:00.000Z protocol upgrade
+        consensus.judgesActivationTime = mainnetConsensus.judgesActivationTime;
 
         // "lrdk" with MSB set
         diskMagic[0] = 0xec;
diff --git a/src/minerfund.cpp b/src/minerfund.cpp
index c46b1dbe35..d2cf31745e 100644
--- a/src/minerfund.cpp
+++ b/src/minerfund.cpp
@@ -60,6 +60,11 @@ std::vector<CTxOut> GetMinerFundRequiredOutputs(const Consensus::Params &params,
         return {};
     }
 
+    if (IsJoshuaEnabled(params, pindexPrev)) {
+        return BuildOutputsCycling(params.coinbasePayoutAddresses.joshua,
+                                   pindexPrev, blockReward);
+    }
+
     if (IsDeuteronomyEnabled(params, pindexPrev)) {
         return BuildOutputsCycling(params.coinbasePayoutAddresses.deuteronomy,
                                    pindexPrev, blockReward);
diff --git a/src/validation.cpp b/src/validation.cpp
index 3c30f7891e..0cf793c29e 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -227,7 +227,7 @@ IsReplayProtectionEnabled(const Consensus::Params &params,
     }
 
     return pindexPrev->GetMedianTimePast() >=
-           activation_time.value_or(params.joshuaActivationTime);
+           activation_time.value_or(params.judgesActivationTime);
 }
 
 /**
diff --git a/test/functional/logos_feature_minerfund_activation.py b/test/functional/logos_feature_minerfund_activation.py
index fd0afbeb1e..969ab884a7 100755
--- a/test/functional/logos_feature_minerfund_activation.py
+++ b/test/functional/logos_feature_minerfund_activation.py
@@ -30,8 +30,9 @@ LEVITICUS_ACTIVATION_TIME = 2010000000
 NUMBERS_ACTIVATION_TIME = 2020000000
 DEUTERONOMY_ACTIVATION_TIME = 2030000000
 JOSHUA_ACTIVATION_TIME = 2040000000
+JUDGES_ACTIVATION_TIME = 2050000000
 
-REPLAYPROTECTION_ACTIVATION_TIME = JOSHUA_ACTIVATION_TIME
+REPLAYPROTECTION_ACTIVATION_TIME = JUDGES_ACTIVATION_TIME
 
 # see consensus/addresses.h, use getaddressinfo to get the scriptPubKey
 GENESIS_SCRIPTS = [
@@ -118,6 +119,9 @@ DEUTERONOMY_SCRIPTS = [
     "76a914017cc7d3bb5512f21835cac23bd92b361393dcd288ac",
     "76a914017cc7d3bb5512f21835cac23bd92b361393dcd288ac",
 ]
+JOSHUA_SCRIPTS = [
+    "76a9142eedc5ad3d142181417daea01076846edfba998088ac",
+]
 
 class MinerFundActivationTest(BitcoinTestFramework):
     def set_test_params(self):
@@ -129,6 +133,7 @@ class MinerFundActivationTest(BitcoinTestFramework):
             f'-leviticusactivationtime={LEVITICUS_ACTIVATION_TIME}',
             f'-numbersactivationtime={NUMBERS_ACTIVATION_TIME}',
             f'-deuteronomyactivationtime={DEUTERONOMY_ACTIVATION_TIME}',
+            f'-joshuaactivationtime={JOSHUA_ACTIVATION_TIME}',
             f'-replayprotectionactivationtime={REPLAYPROTECTION_ACTIVATION_TIME}',
         ]]
 
@@ -309,6 +314,40 @@ class MinerFundActivationTest(BitcoinTestFramework):
             prepare_block(block)
             assert_equal(node.submitblock(ToHex(block)), None)
 
+        # Using the joshua addresses before upgrade fails
+        block = make_block_cb_post_numbers(JOSHUA_SCRIPTS)
+        prepare_block(block)
+        assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+        node.setmocktime(JOSHUA_ACTIVATION_TIME)
+
+        # Mine 11 blocks with JOSHUA_ACTIVATION in the middle
+        # That moves MTP exactly to JOSHUA_ACTIVATION
+        for i in range(-6, 6):
+            block = make_block_cb_post_numbers(DEUTERONOMY_SCRIPTS)
+            block.nTime = JOSHUA_ACTIVATION_TIME + i
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), None)
+
+        assert_equal(node.getblockchaininfo()['mediantime'],
+                     JOSHUA_ACTIVATION_TIME)
+
+        # Now the using the genesis, exodus, leviticus, or numbers, and deuteronomy addresses fails
+        for block in [
+            make_block_with_cb_scripts(GENESIS_SCRIPTS),
+            make_block_with_cb_scripts(EXODUS_SCRIPTS),
+            make_block_with_cb_scripts(LEVITICUS_SCRIPTS),
+            make_block_cb_post_numbers(NUMBERS_SCRIPTS),
+            make_block_cb_post_numbers(DEUTERONOMY_SCRIPTS),
+        ]:
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+
+        # Using the Joshua scripts now works
+        for i in range(0, 12):
+            block = make_block_cb_post_numbers(JOSHUA_SCRIPTS)
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), None)
+
         # Check replay protection is not enabled yet
         tx = CTransaction()
         tx.vin = [CTxIn(COutPoint(int(cointxid, 16), 1))]
